### ISSUE TITLE
Version 9 Interlude forEach. Feeling kind of lost inside functions within functions. Want to make sure I understand this

### ISSUE THREAD

**Issue by ElC-Dev** [gordonmzhu/questions#205]
Hello!
First of all, thank you very much for the great course! I feel like I made a lot of progress so far, and was surprised to see just how so much of the code from the earlier sections stays relevant until the end of the Todo list, and how the whole thing comes together! 

As for my question,It’s not that it I got an error or anything didn’t work as expected. I finished the free course section recently, but After finishing the course, even though I’m sure I improved a lot, I’m still having trouble coding on my own (even simpler stuff than the Todo App in this course), so I decided to rewatch the course to make sure I understood everything concept correctly, as there were some concepts that I just didn’t feel overly confident with. 

There are other things like the concept of  “this” that I’m having a bit of trouble with, but my main weakness right now appears to be how functions inside functions interact, or are enhanced. Even though I’m not sure what exactly is that I’m not getting, I tried my best to pinpoint the video that  I feel best represents my problem. I tried to reexamine the code, and dissect it line by line, as well as use the debugger, and no matter how many times I start thinking “ok, this makes sense”, but after a little time passes, and I re-think about this code, I still can’t help but feel like something is still missing.

To that end, I especially want to make sure I’m interpreting how our own custom forEach from version 9’s interlude works.

### What is the associated Watch and Code lesson? 

 https://watchandcode.com/courses/60264/lectures/1125982
The timestamp is 6:25, where it shows our custom forEach being defined, and its execution.

### Include a live demo link so that I can run your code.  https://glitch.com/edit/#!/honored-waitress?path=script.js:5:2
Not 100% if this a live demo link, but this is the best I've got, sorry.

### Explain the exact steps I need to take to replicate the problem. 

Write your response here (required).
Here’s my interpretation of what happens in  that section of  code:

Basically, forEach is defined with myArray and myFunction as parameters. Then, a For loop is set to loop for as long as i is less than myArray’s length, with i going up by 1 every loop. And each time this happens, myFunction is executed with myArray’s i index as an argument, which resolves to whatever is contained at that array’s i index, and and only this resulting value, that is, whatever is contained inside myArray’s i index is passed on as the 2nd argument when we launch the forEach function. Is this interpretation of the code correct? Any glaring errors or omissions in my logic? I feel like I may be making this sound more convoluted than it should, and despite understanding almost all the todo list code, I still don’t feel too confident about it. 

### What did you expect to happen? 

Write your response here (required).
Everything worked as expected.

### What actually happened? 

Write your response here (required).
There were no errors; the problem I'm having, is that I'm not quite confident if my interpretation of the code is correct. I thought the free section of your was great, but there are still concepts I'm struggling with. Actually, it doesn't even feel like struggling per se, more like I'm able to understand most of the code in the todo list, but a few pieces of the puzzle feel off, and I'm kind of shoehorning them in to make it work. and  It took me quite a while to figure out what exactly I'm having trouble with, but I believe one of them is the way functions inside functions, and how functions are enhanced by this work. Especially in what way parameters are resolved. and are then used as arguments when calling other functions. In the code example above, I think it resolves to something simpler, like just the string coming from the i position in the array's index, but not too long ago, I thought it was a bigger part of the function, and got lost several times, and my overall view of the end result swayed back and forth more times than I care to admit.

### Describe what you've done so far to troubleshoot the problem. 

Write your response here (required).
I re-watched the entire course, and then re-watched videos on functions, forEach again after that. forEach, I think I watched 5 times so far. And I'm still not very good at it, but I also used the debugger several times as well. 

There’s a warning asking us to be completely comfortable with all the concepts and code on Practical Javascript before continuing on to premium, but I honestly don’t feel like I’m quite there yet. While coding along went well for most of the course, and I toyed with some of the code, but when it comes to higher order functions, or functions inside functions,  I just don't feel like I've mastered them. And while re-watching helped a little, it still feels like I'm missing something.

I would like  to make sure I’m really ready, and that I really understood every concept on the free section. If I may ask, is it really okay to move to premium, without feeling fully confident? Is there a way to be relatively sure you’re really ready to move on, or that you actually got most of the basics down?

Sorry for taking so long to get to the point. English isn’t my first language, and even if it was, I’m sure I would still have trouble expressing this issue. 

Thanks for your time!

**Comment by gordonmzhu** [comment]
I don't understand this portion. It's a really long sentence. It might help to break it up and show code examples of what you mean.

```
And each time this happens, myFunction is executed with myArray’s i index as an argument, which resolves to whatever is contained at that array’s i index, and and only this resulting value, that is, whatever is contained inside myArray’s i index is passed on as the 2nd argument when we launch the forEach function. 
```

**Comment by ElC-Dev** [comment]
Thanks for responding!
Sorry for making it confusing. I tried to break it down a bit more, tried to include the lines I believe my description applies to to make it easier to read.  It's the same code I copied from the custom forEach function the version 9 interlude. I included at the live demo link. But should be easier to find in the same comment for this. [https://glitch.com/edit/#!/honored-waitress?path=script.js:5:2]( https://glitch.com/edit/#!/honored-waitress?path=script.js:5:2)

```javascript
var students = [ 'jonathan', 'jenny', 'elliot' ]; 
function ForEach(myArray, myFunction) {
	for (var i = 0; i < myArray.length; i++) {
		myFunction(myArray[i]);
	}
}
```

Line 1: An array with 3 items; all string values, is declared.
Lines 2~6: forEach is defined with myArray and myFunction as parameters. Then, a For loop is set to loop for as long as i is less than myArray’s length, with i going up by 1 every loop. myArray will serve as a container for the whole array, and myFunction will serve as a container for any function we use or define later on.

```javascript
ForEach(students, function(student) {
	console.log(student);
});
```

Lines 7~9: ForEach function is run with students and an anonymous function as it's parameters. The anonymous function defined here uses the student parameter to store data and the anonymous function runs console log with student as it's argument to display the contents of the data stored in the same named parameter defined earlier. 
In this particular case, the data that will be ultimately stored, and output in student by console log will be student's i index. 
For example; in the first loop, the value contained within the student parameter/argument will be just a string called 'jonathan', then a string called jenny on the second loop, and finally a string called elliot on the last loop. There's no additional value; only a single string is stored in this parameter to be called as an argument each time. 

Is this any better? Hopefully my interpretation of the code is correct and I didn't make it even more confusing, instead of simpler I did try to break down as much as I could, but the main reason I'm being this overly detailed describing the code here, is because I'm still having a bit of trouble picturing  exactly what piece of data is being stored in these parameters, before they are invoked as arguments, and want to make sure I'm getting this right.

In my vision of this simple forEach's anonymous function's execution, is that all the loops happening here boil down to outputting a single value stored inside the anonymous function's student parameter, and the single value is just a copy of the same string contained in the i index in the student array.


Also. When explaining code, would it be helpful, or better to edit the glitch link to include comment lines detailing what I believe each line of code is supposed to be doing?

**Comment by ElC-Dev** [comment]
> 
> 
> I don't understand this portion. It's a really long sentence. It might help to break it up and show code examples of what you mean.
> 
> > "And each time this happens, myFunction is executed with myArray’s i index as an argument, which resolves to whatever is contained at that array’s i index, and and only this resulting value, that is, whatever is contained inside myArray’s i index is passed on as the 2nd argument when we launch the forEach function."

Sorry, I guess comments don't count as reply for these. I couldn't find an actual reply button, except for quote reply. Hope this is the correct way to do it for next time. I tried to break down my explanation of the code in that section through a comment,  Please check the comment above for my latest response.
Thanks!

**Comment by gordonmzhu** [comment]
You can inspect these things *directly* by using the debugger. You don't need to wonder about what is happening.

>  I'm still having a bit of trouble picturing exactly what piece of data is being stored in these parameters, before they are invoked as arguments, and want to make sure I'm getting this right.

**Comment by ElC-Dev** [comment]
Thanks. I did use the debugger several times before posting this question, but I guess I still couldn't get it to sink in at the time, and ended up feeling unsure if I really got it, or if my interpretation of what I saw in the debugger was really correct. But after giving it some time, I feel like I'm starting to get a better feel on how code is being interpreted. 

One thing I also noticed while using the debugger, is that it's possible to input the parameter on the console, during debugging to see what is currently stored at that point of code execution. It never occurred me to type variable names during debugging. Hopefully I'll be able to notice stuff like this sooner. Thanks for taking the time to read and respond!

**Comment by gordonmzhu** [comment]
Sounds good. Yeah keep going and if you have trouble interpreting something in debugger, feel free to post a question about that with accompanying screenshots.

**Comment by ElC-Dev** [comment]
Will do. Thank you Gordon :)

**Comment by orientalsniper** [comment]
Hello, I also had trouble understanding fully this lesson, I just didn't understand the need for the "student" parameter for the anonymous function and the similarity of the "student" parameter name with the "students" array got me even more confused. It took me a few days even with the debugger's help to "get it" and I know you explained it in the video.

The reason the "student" parameter is needed is because console.log( ) requires it and the reason it works is because "myArray[i]" gets assigned to the "student" parameter each time the "for loop" runs.

And the reason why the "logName" function didn't require a parameter is because when we created the "logName(name)" function we already created the "name" parameter/argument.
